float32 sign_type
float64 area
